src/commands/subcommand.cc

```diff
@@ -144,37 +136,78 @@ namespace	// subcommand detetction
     return false;
   }
 
-  template <class OutputIterator_>
-  unsigned collectSubommandsIn( const Pathname & dir_r, OutputIterator_ result_r )
+  /** Return whether \a dir_r/name_r forms a valid subcommand. */
+  inline SubcommandOptions::Detected detectSubcommand( const Pathname & dir_r, std::string name_r )
   {
-    unsigned cnt = 0;
+    SubcommandOptions::Detected ret;
+    if ( str::startsWith( name_r, "zypper-" ) && canExecute( dir_r/name_r ) ) {
+      ret._cmd  = name_r.substr( 7 /*"zypper-"*/ );
+      ret._name = std::move(name_r);
+      ret._path = dir_r;
+    }
+    return ret;
+  }
+
+  /** Collect subcommands found in \a dir_r. */
+  inline void detectSubcommandsIn( const Pathname & dir_r, std::function<void(SubcommandOptions::Detected)> fnc_r )
+  {
+    if ( !fnc_r )
+      return;
+
     filesystem::dirForEach( dir_r,
-			    [&result_r]( const Pathname & dir_r, std::string str_r )->bool
+			    [&fnc_r]( const Pathname & dir_r, std::string name_r )->bool
 			    {
-			      if ( str::startsWith( str_r, "zypper-" ) && canExecute( dir_r/str_r ) )
-			      {
-				str_r.erase( 0, 7 /*"zypper-"*/ );
-				*result_r = str_r;
-			      }
+			      SubcommandOptions::Detected cmd { detectSubcommand( dir_r, std::move(name_r) ) };
+			      if ( ! cmd._cmd.empty() )
+				fnc_r( std::move(cmd) );
 			      return true;
 			    } );
-    return cnt;
   }
 
-  inline void collectAllSubommands( std::set<std::string> & execdirCommands_r, std::set<std::string> & pathCommands_r )
+  /* Just the command names for the short help. */
+  inline void collectAllSubommands( std::set<std::string> & allCommands_r )
```

Comment:
collectAllSubommands -> collectAllSubCommands

Reply:
Also its a bit confusing to overload that function later. This is not really collecting all the subcommands but only the names of them. Should the function be renamed to reflect that?

Reply:
collectAllSubCommandNames